packages/next/src/build/static-paths/app.ts

```diff
@@ -1,33 +1,240 @@
 import type { AppSegment } from '../segment-config/app/app-segments'
 import type { StaticPathsResult } from './types'
-import type { Params } from '../../server/request/params'
+import type { CacheHandler } from '../../server/lib/incremental-cache'
 
 import path from 'path'
-import {
-  FallbackMode,
-  fallbackModeToStaticPathsResult,
-} from '../../lib/fallback'
-import * as ciEnvironment from '../../server/ci-info'
-import { formatDynamicImportPath } from '../../lib/format-dynamic-import-path'
-import { interopDefault } from '../../lib/interop-default'
 import { AfterRunner } from '../../server/after/run-with-after'
 import { createWorkStore } from '../../server/async-storage/work-store'
+import { FallbackMode } from '../../lib/fallback'
+import { formatDynamicImportPath } from '../../lib/format-dynamic-import-path'
+import { getRouteMatcher } from '../../shared/lib/router/utils/route-matcher'
+import {
+  getRouteRegex,
+  type RouteRegex,
+} from '../../shared/lib/router/utils/route-regex'
+import { IncrementalCache } from '../../server/lib/incremental-cache'
+import { interopDefault } from '../../lib/interop-default'
 import { nodeFs } from '../../server/lib/node-fs-methods'
-import { getParamKeys } from '../../server/request/fallback-params'
-import { buildStaticPaths } from './pages'
+import { normalizePathname, encodeParam } from './utils'
+import * as ciEnvironment from '../../server/ci-info'
+import escapePathDelimiters from '../../shared/lib/router/utils/escape-path-delimiters'
+
+/**
+ * Compares two parameters to see if they're equal.
+ *
+ * @param a - The first parameter.
+ * @param b - The second parameter.
+ * @returns Whether the parameters are equal.
+ */
+function areParamValuesEqual(a: ParamValue, b: ParamValue) {
+  // If they're equal, then we can return true.
+  if (a === b) {
+    return true
+  }
+
+  // If they're both arrays, then we can return true if they have the same
+  // length and all the items are the same.
+  if (Array.isArray(a) && Array.isArray(b)) {
+    if (a.length !== b.length) {
+      return false
+    }
+
+    return a.every((item) => b.includes(item))
+  }
+
+  // Otherwise, they're not equal.
+  return false
+}
+
+/**
+ * Filters out duplicate parameters from a list of parameters.
+ *
+ * @param paramKeys - The keys of the parameters.
+ * @param routeParams - The list of parameters to filter.
+ * @returns The list of unique parameters.
+ */
+function filterUniqueParams(
+  paramKeys: readonly string[],
+  routeParams: readonly Params[]
+): Params[] {
+  const unique: Params[] = []
+
+  for (const params of routeParams) {
+    let i = 0
+    for (; i < unique.length; i++) {
+      const item = unique[i]
+      let j = 0
+      for (; j < paramKeys.length; j++) {
+        const key = paramKeys[j]
+
+        // If the param is not the same, then we need to break out of the loop.
+        if (!areParamValuesEqual(item[key], params[key])) {
+          break
+        }
+      }
```

Comment:
(non-blocking nit)

this function is a pretty clever usage of loops but I worry that what it makes up for in performance it sacrifices in readability. I'm also not 100% sure that this is more performant than the more expressive version, eg:

```js
 for (const params of routeParams) {
    const isDuplicate = unique.some((item) =>
      paramKeys.every((key) => areParamValuesEqual(item[key], params[key]))
    );

    if (!isDuplicate) {
      unique.push(params);
    }
  }
  ```
  
  In the worst case, this is O(N * M * K), which I think matches the implementation here as well:
  
  ```
  for each params in routeParams (N times):
    for each item in unique (M times):
      for each key in paramKeys (K times):
      // perform comparison
  ```
  
(I actually am curious if the built-in loops that power `some`/`every` might be slightly more efficient?)

Reply:
The only downside to the `.some.every` version is that each loop of `N`, we create 2 functions, which is avoided with the looping. I'll defer to @gnoff if this seems alright 😅

Reply:
I'm partial to the loop code b/c it should be faster both in internal implementation and b/c it creates fewer functions. Typlically for loops can be heavily optimized by JIT and the builtins are only optimized maybe in the newest runtimes and I suspect not likely nearly as much as raw loops. That said I think you can achieve better complexity by tracking the param 'key' becuase the routeParams are in order and spaces aren't allowed.

for /app/[foo]/[bar]/page.tsx
we'd encode
`{}` -> `''`
`{ foo: 'f0' }` -> `'f0'`
`{ bar: 'b1',  foo: 'f1' }` -> `'f1 b1'`
`{ foo: 'f2', bar: 'b2' }` -> `'f2 b2'`
And so if you later encountered `{ bar: 'b2', foo: 'f2' }` (same values, different object key order) you'd just recompute `'f2 b2'` and see you already produced this result

It doesn't even have to be a string you could also have Maps of Maps but this has fewer objects created

The interesting question is what happens if you had just `{ bar: 'b3' }`. This would really be treated like unknown params from the foo downwards so it should encode as `''` since the bar can't be "reached" except if there was a known foo value.